### ISSUE TITLE
Unable to execute addSimpleProduct mutation while other items out of stock

### ISSUE THREAD

**Issue by logaretm** [magento/magento2#26683]
### Summary (*)

We are using Magento as a headless backend for our front-facing application, we are using the GraphQL API to build the frontend.

We noticed if we added a product to cart, and that product later became **out of stock**, any further attempts to add items to cart will fail with `Some cart items are out of stock` effectively locking up the cart from adding other items or changing other items quantities.

This is the returned GraphQL error:

```
"message": "Some of the products are out of stock.",
```

This is a breaking issue for us, as the alternative would result in a very poor experience for our users.

Currently, we have things set up that the frontend is able to detect whenever this exception is thrown and tries to amend it by removing the offending items and attempting to re-play whatever mutations are done to the cart since then.

Needless to say, this is a very flaky solution and it doesn't make sense to have this restriction in the first place since the REST API seems to be fine with this scenario.

### Examples (*)

To Reproduce:

1. Using the GraphQL API: **Add an item to cart.**
2. **Mark it as out-of-stock from the admin dashboard.**
3. Using the GraphQL API: **Try to add other products to cart.**

**I have successfully reproduced this on a fresh magento environment**

### Proposed solution

Ideally, the GraphQL API should allow mutating the cart as long as the out-of-stock item is not being the target of the mutation (updating its quantity for example).

A slightly less robust solution is to allow the mutation to go through, but return `errors` alongside the `data` prompting the GQL client to update their UI and also be aware of possible problems.

### GraphQL requests for reproducing
Create an empty cart
```graphql
mutation {
  createEmptyCart
}
```
Add product to cart 
```graphql
mutation {
  addSimpleProductsToCart(
    input: {
      cart_id: "{ CART_ID }"   # <--- cart id from the previous request goes here
      cart_items: [
        {
          data: {
            quantity: 1
            sku: "simple-product"   # <--- product sku goes here
          }
        }
      ]
    }
  ) {
    cart {
      items {
        id
        product {
          sku
          stock_status
        }
        quantity
      }
    }
  }
```

**Comment by logaretm** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @logaretm. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @logaretm, here is your [Magento instance](https://i-26683-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-26683-2-4-develop.instances.magento-community.engineering/admin_c31c](https://i-26683-2-4-develop.instances.magento-community.engineering/admin_c31c)
Login: `b054e3e7` Password: `656c5038bda6`
Instance will be terminated in up to 3 hours.

**Comment by MeeGz** [comment]
Overcome this issue by overriding Model Cart adding line bolded below to remove error during process.
`<?php
declare(strict_types=1);

namespace Robusta\CartStockErrorHandler\Model\Cart;

use Magento\Framework\Exception\LocalizedException;
use Magento\Framework\GraphQl\Exception\GraphQlInputException;
use Magento\Framework\GraphQl\Exception\GraphQlNoSuchEntityException;
use Magento\Framework\Message\AbstractMessage;
use Magento\Quote\Api\CartRepositoryInterface;
use Magento\Quote\Model\Quote;
use Magento\QuoteGraphQl\Model\Cart\AddProductsToCart as MAddProductsToCart;
use Magento\QuoteGraphQl\Model\Cart\AddSimpleProductToCart;

class AddProductsToCart extends MAddProductsToCart
{
    /**
     * @var CartRepositoryInterface
     */
    private $cartRepository;

    /**
     * @var AddSimpleProductToCart
     */
    private $addProductToCart;

    /**
     * @param CartRepositoryInterface $cartRepository
     * @param AddSimpleProductToCart $addProductToCart
     */
    public function __construct(CartRepositoryInterface $cartRepository, AddSimpleProductToCart $addProductToCart)
    {
        $this->cartRepository = $cartRepository;
        $this->addProductToCart = $addProductToCart;
        parent::__construct($cartRepository, $addProductToCart);
    }

    /**
     * Add products to cart
     *
     * @param Quote $cart
     * @param array $cartItems
     * @throws GraphQlInputException
     * @throws LocalizedException
     * @throws GraphQlNoSuchEntityException
     */
    public function execute(Quote $cart, array $cartItems): void
    {
        foreach ($cartItems as $cartItemData) {
            $this->addProductToCart->execute($cart, $cartItemData);
        }

        **$cart->removeErrorInfosByParams('stock', []);**
        if ($cart->getData('has_error')) {
            throw new GraphQlInputException(
                __('Shopping cart error: %message', ['message' => $this->getCartErrors($cart)])
            );
        }

        $this->cartRepository->save($cart);
    }

    /**
     * Collecting cart errors
     *
     * @param Quote $cart
     * @return string
     */
    private function getCartErrors(Quote $cart): string
    {
        $errorMessages = [];

        /** @var AbstractMessage $error */
        foreach ($cart->getErrors() as $error) {
            $errorMessages[] = $error->getText();
        }

        return implode(PHP_EOL, $errorMessages);
    }
}
`

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-31084` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by lenaorobei** [comment]
This is a complex task that requires framework changes. 

Removing errors is not an option. It leads to the wrong flow when user will see cart errors only on place order step. Errors should be returned under `error` node like it is right now, but cart data should be present under `data` node. This is not supported now but this is the way how this issue should be resolved.

**Comment by magento-engcom-team** [comment]
Hi @logaretm. Thank you for your report. 
The issue has been fixed in magento/magento2#27015 by @AleksLi in 2.4-develop branch
Related commit(s):
 - [77fee571409509a8f4e6c825a03de4eaf3ce3388](https://github.com/magento/magento2/commit/77fee571409509a8f4e6c825a03de4eaf3ce3388)
 - [0a79989737fd30ac4f39e0a24d75b5328d635f2c](https://github.com/magento/magento2/commit/0a79989737fd30ac4f39e0a24d75b5328d635f2c)
 - [af251eb2311a7d477f5c6516195b7032704c755a](https://github.com/magento/magento2/commit/af251eb2311a7d477f5c6516195b7032704c755a)
 - [c47a6a26aae710f2fe64e92d25e03e456c90ae79](https://github.com/magento/magento2/commit/c47a6a26aae710f2fe64e92d25e03e456c90ae79)
 - [39ade1f9791606927a37386d08a96105e4d18613](https://github.com/magento/magento2/commit/39ade1f9791606927a37386d08a96105e4d18613)
 - [2d8721912c321c2098db2e458ec1a0c6c1155658](https://github.com/magento/magento2/commit/2d8721912c321c2098db2e458ec1a0c6c1155658)
 - [158f788ec5a49557104fe1722c7e9551721b5932](https://github.com/magento/magento2/commit/158f788ec5a49557104fe1722c7e9551721b5932)
 - [fff5ce9c09b3d0e9650d7754dff384aad74af7f1](https://github.com/magento/magento2/commit/fff5ce9c09b3d0e9650d7754dff384aad74af7f1)

The fix will be available with the upcoming 2.4.0 release.
